frameless-runtime/src/utxo.rs

```diff
@@ -0,0 +1,253 @@
+            Ok(sig) => sig,
+            Err(_) => return false,
+        };
+        sp_io::crypto::sr25519_verify(&signature, &tx, &Public::from_h256(self))
+    }
+}
+
+// TODO: Implement this for Each Tuxedo Piece
+pub trait UtxoSet {
+
+    /// TODO: Change these bool return types to Result types for more error propagation clarity
+
+    /// Check whether a given utxo exists in the current set
+    fn contains(utxo_ref: UtxoRef) -> bool;
+
+    /// Insert the given utxo into the state storing it with the given ref
+    /// The ref is probably the hash of a tx that created it and its index in that tx, but this decision is opaque to this trait
+    /// Return whether the operation is successful (It can fail if the ref is already present)
+    fn insert(utxo_ref: UtxoRef, utxo: Utxo) -> bool;
+
+    ///
+    /// nullify the utxo by either:
+    /// - Consuming it entirely
+    /// - Putting it on "Timeout"
+    /// - Not consuming it but marking it as spent
+    ///
+    fn nullify(utxo_ref: UtxoRef) -> Option<Utxo>;
+
+    fn peak(utxo_ref: UtxoRef) -> Option<Utxo> {
+        let encoded_utxo = sp_io::storage::get(&utxo_ref.encode())?;
+        match Utxo::decode(&mut &encoded_utxo[..]) {
+            Ok(utxo) => Some(utxo),
+            Err(_) => None,
+        }
+    }
+}
+
+pub trait Get<T> {
+    fn get(&self) -> T;
+}
+
+/// The API of a Tuxedo Piece
+pub trait TuxedoPiece {
+    /// The type of data stored in Outputs associated with this Piece
+    type Data: Encode + Decode;
+    const TYPE_ID: TypeId;
+
+    /// The validation function to determine whether a given input can be consumed.
+    fn validate(transaction: Transaction) -> bool;
+}
+
+pub struct PieceExtracter<Piece>(PhantomData<Piece>);
+impl<Piece: TuxedoPiece> PieceExtracter<Piece> {
+    pub fn extract(key: UtxoRef) -> Result<Piece::Data, ()> {
+        let encoded_utxo = sp_io::storage::get(&key.encode()).ok_or(())?;
+        let utxo = Utxo::decode(&mut &encoded_utxo[..]).map_err(|_| ())?;
+        if utxo.data_id != Piece::TYPE_ID {
+            return Err(())
+        }
+        let piece_data = Piece::Data::decode(&mut &utxo.data[..]).map_err(|_| ())?;
+        Ok(piece_data)
+    }
+}
+
+// User defined logic below for the STF..
+
+pub struct MoneyPiece; // Decodes Value -> u128
+impl UtxoSet for MoneyPiece {
+    fn contains(utxo_ref: UtxoRef) -> bool {
+        sp_io::storage::exists(&utxo_ref.encode())
+    }
+
+    fn insert(utxo_ref: UtxoRef, utxo: Utxo) -> bool {
+        sp_io::storage::set(&utxo_ref.encode(), &utxo.encode());
+        true
+    }
+
+    /// For the Money UTXO we just want to consume it.
+    fn nullify(utxo_ref: UtxoRef) -> Option<Utxo> {
+        let encoded_utxo = sp_io::storage::get(&utxo_ref.encode())?;
+        sp_io::storage::clear(&utxo_ref.encode());
+        match Utxo::decode(&mut &encoded_utxo[..]) {
+            Ok(utxo) => Some(utxo),
+            Err(_) => None,
+        }
+    }
+}
```

Comment:
none of this is money-specific. Makes me think we should not have to impl this trait for each piece.

Reply:
Yes to the second comment and first comment I agree can rework this.